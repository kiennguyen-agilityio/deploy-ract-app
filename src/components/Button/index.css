.btn-primary {
  background-color: rgba(243, 244, 248, 0.65);
  min-width: 150px;
  text-align: center;
  color: var(--color-tertiary);
  border-radius: 4px;
  line-height: var(--line-height-l);
  text-decoration: none;
  border: 2px solid rgba(0, 0, 0, 0.1);
  font-weight: var(--font-weight-regular);
  font-size: var(--font-size-md);
  letter-spacing: 0.5px;
  transition: all 0.2s ease-in-out 0.2s;
  cursor: pointer;
}

.btn-primary:hover {
  background-color: var(--color-primary);
}

.btn-secondary {
  position: relative;
  background-color: transparent;
  color: var(--color-secondary);
  border: 2px solid var(--color-secondary);
  line-height: var(--line-height-l);
  min-width: 150px;
  border-radius: 5px;
  cursor: pointer;
  letter-spacing: 0.5px;
  transition: all 0.2s ease-in-out 0.2s;
  transition: background-color 0.3s ease;
}

.btn-secondary::before {
  content: '+';
  position: absolute;
  left: 0%;
  top: 50%;
  transform: translate(-50%, -50%);
  font-weight: bold;
  color: var(--color-primary);
  background-color: var(--color-secondary);
  width: 30px;
  height: 30px;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.btn-secondary:hover::before {
  color: var(--color-primary);
}

.btn-tertiary {
  background-color: var(--color-secondary);
  color: var(--color-primary);
  border-radius: 4px;
}

.btn.btn-sm {
  padding: 15px 20px;
}

.btn.btn-xl {
  padding: 10px 85px;
}

.btn.btn-lg {
  padding: 10px 63px;
}

.btn.btn-md {
  padding: 10px 35px;
}
