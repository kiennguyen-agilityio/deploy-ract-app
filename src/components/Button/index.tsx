import React, { ReactElement } from 'react';

// styles
import './index.css';

export interface ButtonProps {
  label: string;
  variant: 'primary' | 'secondary' | 'tertiary';
  size?: 'sm' | 'md' | 'lg' | 'xl';
  color?: 'success' | 'warning' | 'default';
  type?: 'button' | 'submit' | 'reset';
  count?: number;
  isDisabled?: boolean;
  isLoading?: boolean;
  onClick?: () => void;
}

const Button = ({
  label,
  variant = 'primary',
  size = 'sm',
  color = 'default',
  type = 'button',
  count = 0,
  isDisabled,
  isLoading,
  onClick,
}: ButtonProps): ReactElement => {
  const classes = `btn btn-${variant} btn-color-${color} btn-${size} ${
    isDisabled ? 'btn-disabled' : ''
  } ${isLoading ? 'btn-loading' : ''}`;

  return (
    <button type={type} className={classes} onClick={onClick} disabled={isDisabled || isLoading}>
      {isLoading && <span className="loader" />}
      <span>
        {label}
        {variant === 'primary' && count !== undefined && ` (${count})`}
      </span>
    </button>
  );
};

export default React.memo(Button);
