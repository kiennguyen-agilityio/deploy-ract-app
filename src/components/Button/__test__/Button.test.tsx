import { render, screen, fireEvent } from '@testing-library/react';
import Button, { ButtonProps } from '@/components/Button/Button';

const setup = (props = {}) => {
  const defaultProps: ButtonProps = {
    label: 'Submit',
    variant: 'primary',
    onClick: jest.fn(),
    isLoading: false,
    ...props,
  };

  render(<Button {...defaultProps} />);
};

describe('Button component', () => {
  test('renders button correctly', () => {
    const label = 'Submit';
    setup();

    const buttonElement = screen.getByRole('button', {
      name: (content) => content.startsWith(label),
    });

    expect(buttonElement).toMatchSnapshot();
  });

  test('renders button with count when variant is primary', () => {
    const label = 'Submit';
    const count = 10;
    const buttonText = `${label} (${count})`;

    setup({ label, variant: 'primary', count });

    const buttonElement = screen.getByText(buttonText);

    expect(buttonElement).toMatchSnapshot();
  });

  test('renders button with secondary variant', () => {
    setup({ variant: 'secondary' });

    const buttonElement = screen.getByText('Submit');
    expect(buttonElement).toMatchSnapshot();
  });

  test('calls onClick handler when clicked', () => {
    const label = 'Submit';
    const handleClick = jest.fn();
    setup({ label, onClick: handleClick });

    const buttonElement = screen.getByRole('button', {
      name: (content) => content.startsWith(label),
    });
    fireEvent.click(buttonElement);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('renders button with loading spinner when isLoading is true', () => {
    setup({ isLoading: true });

    const loaderElement = screen.getByRole('button');
    expect(loaderElement).toBeInTheDocument();
  });
});
