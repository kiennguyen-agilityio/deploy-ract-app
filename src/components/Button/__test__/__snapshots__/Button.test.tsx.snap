// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders button correctly 1`] = `
<button
  class="btn btn-primary btn-color-default btn-sm  "
  type="button"
>
  <span>
    Submit
     (0)
  </span>
</button>
`;

exports[`Button component renders button with count when variant is primary 1`] = `
<span>
  Submit
   (10)
</span>
`;

exports[`Button component renders button with secondary variant 1`] = `
<span>
  Submit
</span>
`;
